nixos/modules/services/networking/bore/bore.nix

```
@@ -0,0 +1,159 @@
+
+      server = mkOption {
+        type = attrsOf (submodule ./server-options.nix);
+        description = "Definition of bore remote proxy servers.";
+        default = { };
+      };
+
+      enableLocalPortCheck = mkEnableOption "local port check" // {
+        default = true;
+        example = false;
+      };
+
+      enableServerPortCheck = mkEnableOption "server port check" // {
+        default = true;
+        example = false;
+      };
+    };
+  };
+
+  config =
+    let
+      getEnabled = s: builtins.filter (value: value.enable) (lib.attrValues s);
+      mkBoreService =
+        kind: v:
+        let
+          isServer = (kind == "server");
+        in
+        {
+          name = "bore-${kind}-${v.name}";
+          value = {
+            description = "bore remote proxy service for ${v.name}";
+            enable = true;
+            after = [
+              "network-online.target"
+              "nss-lookup.target"
+            ];
+
+            requires = [
+              "network-online.target"
+              "nss-lookup.target"
+            ];
+
+            wantedBy = [ "multi-user.target" ];
+
+            environment =
+              if isServer then
+                {
+                  BORE_MIN_PORT = toString v.min-port;
+                  BORE_MAX_PORT = toString v.max-port;
+                }
+              else
+                {
+                  BORE_SERVER = v.to;
+                  BORE_LOCAL_PORT = toString v.local-port;
+                };
+
+            script = ''
+              ${optionalString (v.secretFile != null) ''export BORE_SECRET="$(<${v.secretFile})"''}
+
+              ${lib.getExe' cfg.package "bore"} ${kind} \
+            ''
+            + (
+              if isServer then
+                ''--bind-addr="${v.bind-addr}" --bind-tunnels="${v.bind-tunnels}"''
+              else
+                ''--local-host="${v.local-host}" --port=${toString v.remote-port}''
+            );
+
+            serviceConfig = {
+              Restart = "on-failure";
+              RestartSec = 10;
+            };
+          };
+        };
+    in
+    {
+      systemd.services =
+        (lib.genAttrs' (getEnabled cfg.server) (mkBoreService "server"))
+        // (lib.genAttrs' (getEnabled cfg.local) (mkBoreService "local"));
+
+      assertions =
+        let
+          localTuples = map (v: "${v.local-host}:${toString v.local-port}") (getEnabled cfg.local);
+
+          # For remote tuples, filter out entries where remote-port == 0, since that's a legal default
+          remoteTuples = map (v: "${v.to}:${toString v.remote-port}") (
+            builtins.filter (v: v.remote-port != 0) (getEnabled cfg.local)
+          );
+
+          bindAddrs = map (v: v.bind-addr) (getEnabled cfg.server);
+
+          findDuplicates =
+            list:
+            lib.mapAttrsToList (_: x: builtins.elemAt x 0) (
+              lib.filterAttrs (_: x: (builtins.length x) > 1) (builtins.groupBy toString list)
+            );
```

Personally I find this easier to read as a pipe:


```suggestion
          findDuplicates = list: lib.pipe list [
            (builtins.groupBy toString)
            (builtins.mapAttrs (_: builtins.length))
            (lib.filterAttrs (_: len: len > 1))
            builtins.attrNames
          ];
```